salt/states/x509_v2.py

```diff
@@ -0,0 +1,1685 @@
+                )
+            ):
+                rev_changes["changed"].append(x509util.dec2hex(rev.serial_number))
+
+        for cur_ext in cur.extensions:
+            if _get_extension_for_oid(rev.extensions, cur_ext.oid) is None:
+                # an extension was removed from from the revoked certificate
+                rev_changes["changed"].append(x509util.dec2hex(rev.serial_number))
+
+    for rev in current:
+        # certificate was removed from the CRL, probably because it was outdated anyways
+        if (
+            _get_revoked_certificate_by_serial_number(revoked, rev.serial_number)
+            is None
+        ):
+            rev_changes["removed"].append(x509util.dec2hex(rev.serial_number))
+
+    if any(rev_changes.values()):
+        changes["revocations"] = rev_changes
+
+    ext_changes = _compare_exts(current, builder)
+    if any(ext_changes.values()):
+        changes["extensions"] = ext_changes
+    return changes
+
+
+def _compare_exts(current, builder):
+    def getextname(ext):
+        try:
+            return ext.oid._name
+        except AttributeError:
+            return ext.oid.dotted_string
+
+    added = []
+    changed = []
+    removed = []
+    builder_extensions = cx509.Extensions(_getattr_safe(builder, "_extensions"))
+
+    # iter is unnecessary, but avoids a pylint < 2.13.6 crash
+    for ext in iter(builder_extensions):
+        try:
+            cur_ext = current.extensions.get_extension_for_oid(ext.value.oid)
+            if cur_ext.critical != ext.critical or cur_ext.value != ext.value:
+                changed.append(getextname(ext))
+        except cx509.ExtensionNotFound:
+            added.append(getextname(ext))
+
+    for ext in current.extensions:
+        try:
+            builder_extensions.get_extension_for_oid(ext.value.oid)
+        except cx509.ExtensionNotFound:
+            removed.append(getextname(ext))
+
+    return {"added": added, "changed": changed, "removed": removed}
+
+
+def _compare_ca_chain(current, new):
+    if not len(current) == len(new):
+        return False
+    for i, new_cert in enumerate(new):
+        if new_cert.fingerprint(hashes.SHA256()) != current[i].fingerprint(
+            hashes.SHA256()
+        ):
+            return False
+    return True
+
+
+def _getattr_safe(obj, attr):
+    try:
+        return getattr(obj, attr)
+    except AttributeError as err:
+        # since we cannot get the certificate object without signing,
+        # we need to compare attributes marked as internal. At least
+        # convert possible exceptions into some description
+        raise CommandExecutionError(
+            f"Could not get attribute {attr} from {obj.__class__.__name__}. "
+            "Did the internal API of cryptography change?"
+        ) from err
+
+
+def _compareattr_safe(obj, attr, comp):
+    try:
+        return getattr(obj, attr) == comp
+    except AttributeError:
+        return False
+
+
+def _safe_atomic_write(dst, data, backup):
```

Comment:
There is a atomicfile salt utils that already exists. Can that be used instead?

Reply:
From what I can tell, `salt.utils.atomicfile.atomic_open` does not fit here for two reasons:

a) The temporary file is probably created with the default umask, not `177`. This could introduce brief moments where sensitive data is readable for all users on the system.
b) It does not honor the `backup` setting.